accepted/0000-link-options.md

```diff
@@ -0,0 +1,93 @@
+
+# Detailed design
+
+## Desired behavior
+
+The `yarn link` workflow should mimic publishing a package (ex: `dep`) to npm and then installing it in a dependent (ex: `app`), and keep this constraint while you're making changes to the first package. Concretely, `yarn link` should make it so that when you save a change to `dep`, the resulting state is as if you:
+1. Ran `npm publish` in `dep` (assume that it can clobber an existing version, and that you're publishing to a local registry on just your computer).
+2. Ran `yarn add dep` in `app`.
+
+## Why this behavior is great
+
+This solves several problems that "yarn link" has today:
+
+#### Isolating `node_modules` correctly
+
+You can install `dep` in two different apps without sharing the `node_modules` of `dep`. This is a problem with Electron apps, whose V8 version is different than Node's and uses a different ABI. If you have `node-app` and `electron-app` that both depend on `dep`, the native dependencies of `dep` need to be recompiled separately for each app; `node-app/n_m/dep/n_m` must not be the same as `electron-app/n_m/dep/n_m`.
+
+#### Working on multiple versions
+
+You can be developing multiple different versions of `dep`. Say you have two directories, `dep-1` and `dep-2`, which have your v1 and v2 branches checked out, respectively. With "yarn link" it's not possible to make both of these directories linkable at the same time.
+
+This is a problem when you are developing & testing `dep-1` with `old-app` and `dep-2` with `new-app`. You don't want to be going back and forth between `dep-1` and `dep-2` running "yarn link" each time you switch which app you're testing.
+
+#### Faithfully representing the `node_modules` hierarchy
+
+Currently `yarn link` symlinks the entire package directory, which brings along its `node_modules` subdirectory with it. With dependency deduping and flattening, bringing in `dep/node_modules` wholesale usually produces a different `node_modules` hierarchy than running `yarn install` in `app` and installing everything from npm. This isn't a problem most of the time but it does go against Yarn's spirit of consistency and the lockfile.
+
+## A practical proposal
+
+This is a proposal that solves all of the problems above and isn't too hard to implement or understand. Conceptually, we find all the files we'd normally publish to npm, pack them up using symlinks instead of copies of the files, publish the pack to a local registry (just a directory), and then when installing we look up packages in the local registry directory instead of npm.
+
+### Running "yarn link --pack" inside of `dep`
+
+This is the step that simulates publishing `dep`. Running `yarn link --pack` in `dep` finds all the files that "yarn publish" would pack up and upload to npm. Crucially, this excludes `node_modules`, and would follow the same algorithm as "yarn publish" such as reading package.json's `files` field.

```

Comment:
My interpretation of `yarn knit` goal is to have end to end simulation of consuming a first party package as if it was published to a registry, but without the network part and maybe without wasting CPU on gzipping.

I still think Yarn needs this feature because it enables plenty of use cases without having limits by OS, file system and package folder structure.

I recognize that some people don't want to pay the fs.copy performance tax and would compromise on consistency by using `yarn link` or `link:/` as long as there is a way to blacklist `node_modules` and other dev folders.
That is why I think the proposed solution should be separate from the `knit` proposal.

As for "compromise on consistency" I am not convinced that transforming
```
~/work/project_a/node_modules/dep -> (symlink) ../../dep
```
into
```
~/work/project_a/node_modules/dep -> <buffer folder> -> (symlink) ../../dep/file1
~/work/project_a/node_modules/dep -> <buffer folder> -> (symlink) ../../dep/file2
~/work/project_a/node_modules/dep -> <buffer folder> -> (symlink) ../../dep/file3
~/work/project_a/node_modules/dep -> <buffer folder> -> (symlink) ../../dep/file4
~/work/project_a/node_modules/dep -> <buffer folder> -> (symlink) ../../dep/folder1
~/work/project_a/node_modules/dep -> <buffer folder> -> (symlink) ../../dep/folder2
~/work/project_a/node_modules/dep -> <buffer folder> -> (symlink) ../../dep/folder3/file1
...
```
would be easy to implement and maintain.

It feels like it will leave clutter and stale references when used heavily.
That is why I want it to be localized where it is easy to clear and monitor, node_modules of the project seems reasonable.

What would be the reason to make it global?